### PR TITLE
refactor: `_build_trials_seq`

### PR DESCRIPTION
**Pull request by DaniBodor**
fix: #29

- refactor: `_build_trials_seq` without `max_sequential` 
- refactor: tests associated to above
- test: fixed bug in `test_build_trials_seq_maximum_sequential_trials`
- fix: max constrained shuffle
- docs: update `Task` docstring and code comments
- refactor: sum _session values to 1

### CODE REVIEW

**Comment by DaniBodor** [comment]
@gcroci2 , I now fixed the bugs that I introduced, so it's ready to review. You can do it one commit at a time, there is little to no overlap between them and they follow a logical order.

File: tests/test_task.py [hunk 1]
```diff
@@ -238,26 +238,25 @@ def test_build_trials_seq_distributions(session: dict, catch_prob: float):
...
+    counts = {modality: np.sum(task._modality_seq == modality) for modality in [*task._modalities, "catch"]}
+    tolerance = 0.2
+    for i, comp in enumerate(
+        [
+            (counts["catch"] / len(task._modality_seq), task.catch_prob),
+            (counts["v"] / len(task._modality_seq), task.session["v"] - task.catch_prob * task.session["v"]),
+            (counts["a"] / len(task._modality_seq), task.session["a"] - task.catch_prob * task.session["a"]),
+        ],
+    ):
+        assert np.isclose(
```
Comment (gcroci2) [inline tests/test_task.py:252]: This is more compact than before, but very hard to follow. I'd revert it to three separate asserts, or use clearer variables names (e.g., for `comp`; but I don't have nice ideas about how to improve that, and I wouldn't spend time in doing so)

File: tests/test_task.py [hunk 2]
```diff
@@ -267,19 +266,32 @@ def test_build_trials_seq_shuffling():
     _ = task_shuffled.generate_trials(ntrials=NTRIALS)
     _ = task_not_shuffled.generate_trials(ntrials=NTRIALS)
 
-    # Verify that the generated sequences are shuffled or not shuffled accordingly
     assert task_shuffled._modality_seq.shape == task_not_shuffled._modality_seq.shape
-    assert not np.array_equal(task_shuffled._modality_seq, task_not_shuffled._modality_seq)
+
+    keys = list(task_shuffled._session.keys())
+    first_occurrence = list(task_not_shuffled._modality_seq).index(keys[1])
+    assert (
```
Comment (gcroci2) [inline tests/test_task.py:273]: In which cases is this going to be False? I didn't get the point of this assert

File: tests/test_task.py [hunk 3]
```diff
@@ -267,19 +266,32 @@ def test_build_trials_seq_shuffling():
...
 def test_build_trials_seq_maximum_sequential_trials():
     # Create a Task instance with shuffling enabled and a maximum sequential trial constraint
     task = Task(name=NAME, max_sequential=4)
     _ = task.generate_trials(ntrials=NTRIALS)
+
     # Ensure that no more than the specified maximum number of consecutive trials of the same modality occur
-    for modality in task._modalities:
-        for i in range(len(task._modality_seq) - task.max_sequential):
-            assert np.sum(task._modality_seq[i : i + task.max_sequential] == modality) <= task.max_sequential
+    sequence_string = "".join(task._modality_seq).replace("catch", "X")
```
Comment (gcroci2) [inline tests/test_task.py:289]: Why is the replace needed?

File: tests/test_task.py [hunk 4]
```diff
@@ -267,19 +266,32 @@ def test_build_trials_seq_shuffling():
...
+    too_many = task.max_sequential + 1
+    for mod in set(sequence_string):
```
Comment (gcroci2) [inline tests/test_task.py:291]: Very nice improvement!!

**Review by gcroci2 — CHANGES_REQUESTED** [review]
Great changes! Minor comments, mainly about code readability :)

I re-requested my own review by mistake

File: annubes/task.py [hunk 5]
```diff
@@ -352,48 +353,54 @@ def _check_time_vars(self) -> None:
...
-                            f"`max_sequential` limit of {self.max_sequential} was violated {violations} times"
-                            f" for {mod}.\n"
-                            "Please check the current trials sequence, and if it is not acceptable redefine your"
-                            f"task or try to change the random seed, now set to {self._random_seed}.\n",
-                            stacklevel=2,
-                        )
+        # Extract options and probs from the dictionary
+        options = list(self._session.keys())
+        probs = list(self._session.values())
+        max_seq = self.max_sequential
```
[the remark this answers is not in the document]
Comment (DaniBodor) [reply annubes/task.py:359]: fair, I thought the shorter lines would make it more readable (similar to we do in line 357-358, which also isn't strictly needed), but I guess this one doesn't really make a big difference.

File: tests/test_task.py (hunk 1 above)
Comment (DaniBodor) [reply tests/test_task.py:255]: I actually found it hard to follow before already. Is this better: addd4bb?
Also, the fact that we need to change the absolute tolerance to 0.2 to keep it from failing regularly makes me wonder whether this test still serves a purpose.

File: tests/test_task.py (hunk 2 above)
Comment (DaniBodor) [reply tests/test_task.py:277]: It's false if there's a "v" after an "a". I added a comment to the code to make this clearer.

File: tests/test_task.py (hunk 3 above)
Comment (DaniBodor) [reply tests/test_task.py:293]: Otherwise `set(sequence_string)` would add each letter of "catch" as a separate test case. There's probably other ways around this, but this seemed simple and sensible to me.

In fact, I was thinking whether we want to replace "catch" by "X" in session (and everywhere) as well. I think given that we say that each modularity is given by a single character, it makes sense to apply this to catches as well, and reserve X (or underscore?) for that purpose.

Comment (DaniBodor) [reply tests/test_task.py:293]: Otherwise `set(sequence_string)` would add each letter of "catch" as a separate test case. There's probably other ways around this, but this seemed simple and sensible to me.

In fact, I was thinking whether we want to replace "catch" by "X" in session (and everywhere) as well. I think given that we say that each modularity is given by a single character, it makes sense to apply this to catches as well, and reserve X (or underscore?) for that purpose.

File: annubes/task.py [hunk 6]
```diff
@@ -352,48 +353,54 @@ def _check_time_vars(self) -> None:
...
+
+        if not (self.shuffle_trials and max_seq):
```
[the remark this answers is not in the document]
Comment (gcroci2) [reply annubes/task.py]: Indeed clear the list would be the way, but I do agree that feels weird. I think it feels already better now with the `self._constrained_shuffle` attribute, but I'd swap the if else statement: instead of starting with the `if not` condition, I'd place the shuffle + max_seq first, and then the else.

File: tests/test_task.py [hunk 7]
```diff
@@ -238,26 +238,29 @@ def test_build_trials_seq_distributions(session: dict, catch_prob: float):
...
+        "catch": task.catch_prob,
+    }
+    tolerance = 0.2
+
+    for modularity, actual_ratio in ratios.items():
+        assert np.isclose(
+            actual_ratio,
+            expected_ratios[modularity],
+            atol=tolerance,
+        ), f"Actual difference for {modularity}: {round(abs(actual_ratio[0] - expected_ratios[modularity]), 3)}"
```
Comment (gcroci2) [inline tests/test_task.py:259]:
```suggestion
        ), f"Actual difference for {modularity}: {round(abs(actual_ratio - expected_ratios[modularity]), 3)}"
```

Comment (gcroci2) [reply tests/test_task.py:259]: Remove the indexing, otherwise the failed assert print does not work

File: tests/test_task.py (hunk 1 above)
Comment (gcroci2) [reply tests/test_task.py:255]: Better :) I think 0.2 tolerance still serves the purpose

File: tests/test_task.py (hunk 3 above)
Comment (gcroci2) [reply tests/test_task.py:290]: I like this! Let's change it everywhere in the code. I prefer `X` to the underscore. I am in favor of doing it in this PR, but in case you prefer feel free to open a new issue for this.

**Review by gcroci2 — CHANGES_REQUESTED** [review]
Almost there! Maaaany improvements in this PR :)

[Note that static lint check is failing]

File: annubes/task.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DaniBodor) [reply annubes/task.py]: For readability, I think it makes more sense to leave it as it is, despite the negative conditional.

Now the order (in the code base) is:
1. unshuffled
2. simple shuffle
3. constrained shuffle.

If I would turn the condition around the order would be: 
1. constrained shuffle
2. unshuffled
3. simple shuffle.

File: tests/test_task.py (hunk 3 above)
Comment (DaniBodor) [reply tests/test_task.py]: Done

File: tests/test_task.py [hunk 8]
```diff
@@ -238,26 +238,29 @@ def test_build_trials_seq_distributions(session: dict, catch_prob: float):
...
+        for modality in [*task._modalities, "catch"]
+    }
+    expected_ratios = {
+        "v": task.session["v"] - task.catch_prob * task.session["v"],
+        "a": task.session["a"] - task.catch_prob * task.session["a"],
+        "catch": task.catch_prob,
+    }
+    tolerance = 0.2
+
+    for modularity, actual_ratio in ratios.items():
```
[the remark this answers is not in the document]
Comment (DaniBodor) [reply tests/test_task.py]: You're right :)

File: annubes/task.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (gcroci2) [reply annubes/task.py]: Ok let's keep it this way

**Review by gcroci2 — APPROVED** [review]
Thanks for the edits :) I haven't checked in depth the new logic for the redraw part, but I trust you that is correct, and all tests pass. As a general note, I think that especially for side features like this one (which now is very nicely correctly implemented, of course), we should try to keep the review process shorter, and code readability easier, at the expenses of less optimal solutions. Hours are limited, and we still need to implement a huge part of the pipeline, so let's try to find a more balanced trade off ;) and if necessary for next times let's discuss together how long we want a certain feature edit/dev to take.
